h2/src/main/org/h2/command/Parser.java

```diff
@@ -5199,149 +5205,156 @@ else if(readIf("DELETE")) {
+        int id = db.allocateObjectId();
+        Column[] columnTemplateArray = columnTemplateList.toArray(new Column[0]);
+        
         // No easy way to determine if this is a recursive query up front, so we just compile
         // it twice - once without the flag set, and if we didn't see a recursive term,
         // then we just compile it again.
-        TableView view = new TableView(schema, id, tempViewName, querySQL,
-                parameters, columnTemplateList.toArray(new Column[0]), session,
-                allowRecursiveQueryDetection, false);
-        if (!view.isRecursiveQueryDetected() && allowRecursiveQueryDetection) {
-            session.removeLocalTempTable(view);
-            view = new TableView(schema, id, tempViewName, querySQL, parameters,
-                    columnTemplateList.toArray(new Column[0]), session,
-                    false/* recursive */, false);
+        TableView view;
+        synchronized(targetSession){            
+            view = new TableView(schema, id, cteViewName, querySQL,
+                    parameters, columnTemplateArray, targetSession,
+                    allowRecursiveQueryDetection, false /* literalsChecked */, true /* isTableExpression */, isPersistent);
+            if (!view.isRecursiveQueryDetected() && allowRecursiveQueryDetection) {
+                if(isPersistent){
+                    db.addSchemaObject(targetSession, view);
+                    view.lock(targetSession, true, true);
+                    targetSession.getDatabase().removeSchemaObject(targetSession, view);                
+                }else{
+                    session.removeLocalTempTable(view);                    
+                }
+                view = new TableView(schema, id, cteViewName, querySQL, parameters,
+                        columnTemplateArray, targetSession,
+                        false/* assume recursive */, false /* literalsChecked */, true /* isTableExpression */, isPersistent);
+            }
+            // both removeSchemaObject and removeLocalTempTable hold meta locks
+            targetSession.getDatabase().unlockMeta(targetSession);            
         }
         view.setTableExpression(true);
-        view.setTemporary(true);
+        view.setTemporary(!isPersistent);
         view.setHidden(true);
+        view.setOnCommitDrop(false);
         if(addViewToSession){
-            session.addLocalTempTable(view);
+            if(isPersistent){
+                db.addSchemaObject(targetSession, view);
+                view.unlock(targetSession);
+                db.unlockMeta(targetSession);
+            }
+            else{
+                targetSession.addLocalTempTable(view);
+            }
         }
-        view.setOnCommitDrop(false);
         return view;
     }
 
     private CreateView parseCreateView(boolean force, boolean orReplace) {
         boolean ifNotExists = readIfNotExists();
+        boolean isTableExpression = readIf("TABLE_EXPRESSION");
```

Comment:
Looks like this is some weird non-standard SQL?

Reply:
I'd prefer to avoid any non-standard SQL extensions. As far as I understand recursive views already were supported without this keyword, no?

Reply:
Also, I do not see tests for it.

Reply:
I don't have a problem with this - we create our own internal SQL in other places too, to allow us to persist extra information about schema objects.

Reply:
Ok, I see. Having a test for internal SQL is still a good idea though.

Reply:
The tests for creating, persisting and re-loading the internal SQL is in all the TestPersistentCommonTableExpressions test cases - after all how best else to make sure the  views persist between sessions. In fact, it was those (user) test cases that highlighted the problem in the first place.